sdk/eventhub/Azure.Messaging.EventHubs/src/Authorization/SharedAccessSignature.cs

```diff
@@ -0,0 +1,435 @@
+        /// <param name="host">The fully qualified host name for the Event Hubs namespace.  This is likely to be similar to <c>{yournamespace}.servicebus.windows.net</c>.</param>
+        /// <param name="eventHubPath">The path of the specific Event Hub to connect the client to.</param>
+        /// <param name="sharedAccessKeyName">The name of the shared access key that the signature should be based on.</param>
+        /// <param name="sharedAccessKey">The value of the shared access key for the signagure.</param>
+        /// <param name="signatureValidityDuration">The duration that the signature should be considered valid.</param>
+        ///
+        public SharedAccessSignature(ConnectionType connectionType,
+                                     string         host,
+                                     string         eventHubPath,
+                                     string         sharedAccessKeyName,
+                                     string         sharedAccessKey,
+                                     TimeSpan       signatureValidityDuration)
+        {
+            Guard.ArgumentNotNullOrEmpty(nameof(host), host);
+            Guard.ArgumentNotNullOrEmpty(nameof(eventHubPath), eventHubPath);
+            Guard.ArgumentNotNullOrEmpty(nameof(sharedAccessKeyName), sharedAccessKeyName);
+            Guard.ArgumentNotNullOrEmpty(nameof(sharedAccessKey), sharedAccessKey);
+
+            Guard.ArgumentNotTooLong(nameof(sharedAccessKeyName), sharedAccessKeyName, MaximumKeyNameLength);
+            Guard.ArgumentNotTooLong(nameof(sharedAccessKey), sharedAccessKey, MaximumKeyLength);
+            Guard.ArgumentNotNegative(nameof(signatureValidityDuration), signatureValidityDuration);
+
+            SharedAccessKeyName = sharedAccessKeyName;
+            SharedAccessKey = sharedAccessKey;
+            ExpirationUtc = DateTime.UtcNow.Add(signatureValidityDuration);
+            Resource = BuildAudience(connectionType, host, eventHubPath);
+            Value = BuildSignature(Resource, sharedAccessKeyName, sharedAccessKey, ExpirationUtc);
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="SharedAccessSignature"/> class.
+        /// </summary>
+        ///
+        /// <param name="sharedAccessSignature">The shared access signature that will be parsed as the basis of this instance.</param>
+        /// <param name="sharedAccessKey">The value of the shared access key for the signature.</param>
+        ///
+        public SharedAccessSignature(string sharedAccessSignature,
+                                     string sharedAccessKey)
+        {
+            Guard.ArgumentNotNullOrEmpty(nameof(sharedAccessSignature), sharedAccessSignature);
+            Guard.ArgumentNotTooLong(nameof(sharedAccessKey), sharedAccessKey, MaximumKeyLength);
+
+            (SharedAccessKeyName, Resource, ExpirationUtc) = ParseSignature(sharedAccessSignature);
```

Comment:
Try to avoid virtual member calls from ctor, the object would not be completely initialized when the method is invoked.

Reply:
Good catch.   I think, with respect to your earlier comment, that I can probably remove `virtual`; it was added for testing purposes but I don't believe that I ended up using it.